### ISSUE TITLE
Strange latency increasing on the quantized model

### ISSUE THREAD

**Issue by MartaSamoilenkoPn** [huggingface/optimum-benchmark#214]
Hello!
I am testing the mistral-7b inference after quantization. I also want to test the impact of Flash Attention (sdpa, eager, fa2) on model inference. But the model decode latency is too high, and it is strange for me. I rent the virtual GPU Nvidia 4090 on Vast.ai).  I am changing the sequence length from 512 to 4096 and have the next results:
![image](https://github.com/huggingface/optimum-benchmark/assets/116800440/e0c5aca4-0d23-4655-bbd1-460ad5dff69f)
![image](https://github.com/huggingface/optimum-benchmark/assets/116800440/e8408006-8c02-4984-923c-2fefbd3e3c56)
![image](https://github.com/huggingface/optimum-benchmark/assets/116800440/d59d77ef-b154-4e5a-9acf-6765a5cc1562)

The code for the benchmarks (I change the models path and quantization/fa schemas depending on the benchmark):

```py
from optimum_benchmark.backends.pytorch.config import PyTorchConfig
from optimum_benchmark.benchmarks.inference.config import InferenceConfig
from optimum_benchmark.experiment import ExperimentConfig, launch
from optimum_benchmark.launchers.process.config import ProcessConfig
from optimum_benchmark.logging_utils import setup_logging
import torch
import os

import json


def benchmark_mistral7b(input_seqlen=1024):

    setup_logging(level="INFO")
    launcher_config = ProcessConfig(device_isolation=False)

    # batch_size = int(16384 / input_seqlen)
    batch_size = 1 

    for i in [512, 1024, 2048, 4096, 8192]:

        print("seqlen: ", i)

        benchmark_config = InferenceConfig(
            memory=True,
            latency=True,
            input_shapes={"batch_size": batch_size, "sequence_length": i},
            generate_kwargs={"max_new_tokens": 512, "min_new_tokens": 512},
        )

        backend_config = PyTorchConfig(
            model="MartaSamoilenko/Mistral-7b-int4-quantized-AWQ",
            device="cuda",
            device_ids="0",
            # torch_dtype = "float16",
            quantization_scheme="awq",
            attn_implementation = "sdpa",
            no_weights=False,
        )
        experiment_config = ExperimentConfig(
            experiment_name="mistral-awq-fa2-base",
            benchmark=benchmark_config,
            launcher=launcher_config,
            backend=backend_config,
        )
        benchmark_report = launch(experiment_config)

        benchmark_report = benchmark_report.to_dict()
        experiment_config = experiment_config.to_dict()

        with open(f"mistral7b_awq_sdpa_seqlen_{i}_batch_size_{batch_size}_benchmark_report.json", "w") as outfile: 
            json.dump(benchmark_report, outfile)
        
        with open(f"mistral7b_awq_sdpa_seqlen_{i}_batch_size_{batch_size}_experiment_config.json", "w") as outfile: 
            json.dump(experiment_config, outfile)


        # benchmark_report.to_json(, flat=False)
        # experiment_config.to_json(, flat=False)



if __name__ == "__main__":
    os.environ["PROCESS_SPECIFIC_VRAM"] = "0"
    benchmark_mistral7b(input_seqlen=16384)
```
If you know what the issue is, please write it down below. Thank you!

**Comment by IlyasMoutawwakil** [comment]
Hi!
- increasing the sequence length wouldn't impact decode latency (as you can see), as decoding uses kv cache so it's like a prefill with seq len = 1 every decoding step (not exactly but most of the matmuls are similar). The impacted metric here would be prefill latency. And know that with flash attention v2, decoding is usually slower, because the kernel is optimized for prefill (long seq len passes), there's another kernel [described here](https://github.com/Dao-AILab/flash-attention?tab=readme-ov-file#22-optimize-for-inference) that's better for decoding but it's still not part of transformers.
- we report multiple latency metrics: prefill, decode and per token, and for each we have mean, p50, p90, stddev, etc. i suggest you either plot p50 (median) or plot the mean with stddev (boxplots) to se how much variance we're talking about. If the stddev is big (more than 5% of the mean) I suggest you add more warmup steps and/or iterations/duration.
- Decode latency as a metric is not very important, since it's cumulative and depend on the `max_new_tokens`/`min_new_tokens`, although I see there's some issue with it in your benchmark (why does it increase a lot with some attn implementations ?). I would say you probably ran the benchmarks concurrently or there's some code that's using the GPU at the same time, enabling device isolation with `device_isolation_action=warn` should give you a hint if that's the case.
- You can make you benchmarks take less time even while increasing warmup steps / iterations, with smaller `max_new_tokens`/`min_new_tokens` but looking at the per token latency, with backends like pytorch we can access fine graoned information about how long each token took, so with `max_new_tokens`/`min_new_tokens`=32, we already get 32 measurement per generation pass, ie with 10 iterations you get 320 and looking at p50 you generally get a good estimate (a good estimate for me is when p50 and p90 are not too far from each other)
- please use the latest version of optimum (0.2.1 on pypi or main), the Experiment schema was deprecated a couple months ago, and the Benchmark schema is more stable.

**Comment by IlyasMoutawwakil** [comment]
okay looking at the llm perf leaderboard I'm seeing the same behavior, I'll try to compare with some handwritten benchmarks.

**Comment by IlyasMoutawwakil** [comment]
tldr; you're on torch 2.3 and custom kernels are failing silently (sometimes 🥲)
- For AWQ: the default layer is GEMM, but the latest awq kernels were compiled with torch 2.2, making them unusable with torch 2.3 (we call it the missing symbols issue 🥲), so instead of failing, the layer silently switch to using a pure pytorch based implementation ; https://github.com/casper-hansen/AutoAWQ/blob/main/awq/modules/linear/gemm.py#L49
- For GPTQ: it's even crazier, because even if you specify ExllamaV2 as your layer with the configuration:
  ```python
  quant_config = GPTQConfig(bits=4, use_exllama=True, exllama_config={"version": 2})
  llama = AutoModelForCausalLM.from_pretrained(
      model_id,
      quantization_config=quant_config,
      device_map=torch.device("cuda"),
  )
  ```
  you'll get different layers in your model depending on whether you're on torch 2.2 or torch 2.3
  ```python
  print(llama.model.layers[0].self_attn.v_proj.__class__)
  <class 'auto_gptq.nn_modules.qlinear.qlinear_exllamav2.QuantLinear'>
  ```
  ```python
  print(llama.model.layers[0].self_attn.v_proj.__class__)
  <class 'auto_gptq.nn_modules.qlinear.qlinear_cuda.QuantLinear'>
  ```
  This can be traced back to this indeterministic function 🥲 https://github.com/AutoGPTQ/AutoGPTQ/blob/main/auto_gptq/utils/import_utils.py#L96

Check the results of switching between torch versions in this benchmark I just uploaded:
https://huggingface.co/datasets/optimum-benchmark/pytorch-llama/tree/main/benchmarks

**Comment by MartaSamoilenkoPn** [comment]
Thank you very much, it helped! I had torch 2.3 :(
